engine/executor/dm/unitholder.go

```diff
@@ -89,105 +68,142 @@ func newUnitHolderImpl(workerType lib.WorkerType, cfg *dmconfig.SubTaskConfig) *
+	}()
 	return nil
 }
 
-func filterErrors(r *pb.ProcessResult) {
-	errs := make([]*pb.ProcessError, 0, 2)
-	for _, err := range r.Errors {
-		if !unit.IsCtxCanceledProcessErr(err) {
-			errs = append(errs, err)
-		}
-	}
-	r.Errors = errs
-}
-
-func (u *unitHolderImpl) fetchAndHandleResult() {
-	if r := u.getResult(); r != nil {
-		return
-	}
+func (u *unitHolderImpl) Pause(ctx context.Context) error {
+	u.processMu.Lock()
+	defer u.processMu.Unlock()
 
-	select {
-	case r := <-u.resultCh:
-		filterErrors(&r)
-		if len(r.Errors) > 0 {
-			// TODO: refactor unit.Syncer
-			// unit needs to manage its own life cycle
-			u.unit.Pause()
-		}
-		u.setResult(&r)
-	default:
+	stage, _ := u.Stage()
+	if stage != metadata.StageRunning && stage != metadata.StageError {
+		return errors.Errorf("failed to pause unit with stage %d", stage)
 	}
-}
 
-func (u *unitHolderImpl) setResult(r *pb.ProcessResult) {
-	u.mu.Lock()
-	defer u.mu.Unlock()
-	u.lastResult = r
+	// cancel process
+	u.fieldMu.Lock()
+	u.runCancel()
+	u.fieldMu.Unlock()
+	u.processWg.Wait()
+	// TODO: refactor unit.Syncer
+	// unit needs to manage its own life cycle
+	u.unit.Pause()
+	return nil
 }
 
-func (u *unitHolderImpl) getResult() *pb.ProcessResult {
-	u.mu.RLock()
-	defer u.mu.RUnlock()
-	return u.lastResult
-}
+func (u *unitHolderImpl) Resume(ctx context.Context) error {
+	u.processMu.Lock()
+	defer u.processMu.Unlock()
 
-func (u *unitHolderImpl) checkAndAutoResume() {
-	result := u.getResult()
-	if result == nil || len(result.Errors) == 0 {
-		return
+	stage, _ := u.Stage()
+	if stage != metadata.StagePaused && stage != metadata.StageError {
+		return errors.Errorf("failed to resume unit with stage %d", stage)
 	}
 
-	log.L().Error("task runs with error", zap.Any("error msg", result.Errors))
-	subtaskStage := &pb.SubTaskStatus{
-		Stage:  pb.Stage_Paused,
-		Result: result,
-	}
-	strategy := u.autoResume.CheckResumeSubtask(subtaskStage, config.DefaultBackoffRollback)
-	log.L().Info("got auto resume strategy", zap.Stringer("strategy", strategy))
-
-	if strategy == worker.ResumeDispatch {
-		log.L().Info("dispatch auto resume task")
-		u.setResult(nil)
-		// TODO: manage goroutines
-		go u.unit.Resume(u.ctx, u.resultCh)
-	}
-}
-
-// Tick implement UnitHolder.Tick
-func (u *unitHolderImpl) Tick(ctx context.Context) error {
-	u.fetchAndHandleResult()
-	u.checkAndAutoResume()
+	runCtx, runCancel := context.WithCancel(context.Background())
+	// run new process
+	u.fieldMu.Lock()
+	u.runCtx, u.runCancel = runCtx, runCancel
+	u.result = nil
+	u.fieldMu.Unlock()
+
+	resultCh := make(chan pb.ProcessResult, 1)
+	u.processWg.Add(1)
+	go func() {
+		defer u.processWg.Done()
+		u.unit.Resume(runCtx, resultCh)
+		u.fetchAndHandleResult(resultCh)
+	}()
 	return nil
 }
 
 // Close implement UnitHolder.Close
 func (u *unitHolderImpl) Close(ctx context.Context) error {
-	u.cancel()
+	u.processMu.Lock()
+	defer u.processMu.Unlock()
+
+	u.fieldMu.Lock()
+	// cancel process
+	u.runCancel()
```

Comment:
will framework call CloseImpl before or concurrent to InitImpl?

Reply:
No